segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ReadOnlySegmentContainer.java

```diff
@@ -128,14 +130,20 @@ public boolean isOffline() {
     public CompletableFuture<ReadResult> read(String streamSegmentName, long offset, int maxLength, Duration timeout) {
         Exceptions.checkNotClosed(this.closed.get(), this);
         TimeoutTimer timer = new TimeoutTimer(timeout);
-        return getStreamSegmentInfo(streamSegmentName, false, timer.getRemaining())
-                .thenApply(si -> StreamSegmentStorageReader.read(si, offset, maxLength, MAX_READ_AT_ONCE_BYTES, this.storage));
+        return Retry.withExpBackoff(30, 10, 4)
+                    .retryingOn(StreamSegmentNotExistsException.class)
+                    .throwingOn(RuntimeException.class)
+                    .run(() -> getStreamSegmentInfo(streamSegmentName, false, timer.getRemaining())
```

Comment:
You can move lines 133-135 to a private static final field, and just reference it from here